packages/ampd-sdk/src/grpc/connection_pool.rs

```diff
@@ -0,0 +1,394 @@
+use std::mem::discriminant;
+
+use ampd::url::Url;
+use error_stack::Result;
+use report::ResultExt;
+use thiserror::Error;
+use tokio::sync::{mpsc, watch};
+use tokio::time::{timeout, Duration};
+use tokio_util::sync::CancellationToken;
+use tonic::transport;
+use tracing::{error, info, warn};
+
+use crate::future::{with_retry, RetryPolicy};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to connect to the grpc endpoint")]
+    GrpcConnection(#[from] transport::Error),
+
+    #[error(transparent)]
+    ClientNotificationFailed(#[from] watch::error::SendError<ConnectionState>),
+
+    #[error(transparent)]
+    SelfNotificationFailed(#[from] mpsc::error::SendError<ClientMessage>),
+
+    #[error(transparent)]
+    ConnectionStateReceiveFailed(#[from] watch::error::RecvError),
+
+    #[error("invalid url")]
+    InvalidUrl,
+
+    #[error("connection pool disconnected")]
+    ServerDisconnected(#[from] tokio::time::error::Elapsed),
+}
+
+// TODO: make these configurable
+const DEFAULT_RETRY_POLICY: RetryPolicy = RetryPolicy::RepeatConstant {
+    sleep: Duration::from_millis(100),
+    max_attempts: 3,
+};
+const DEFAULT_SERVER_TIMEOUT: Duration = Duration::from_secs(30);
+const DEFAULT_INITIAL_TIMEOUT: Duration = Duration::from_secs(3);
+const DEFAULT_RPC_TIMEOUT: Duration = Duration::from_secs(5);
+const KEEPALIVE_TIME: Duration = Duration::from_secs(30);
+const KEEPALIVE_TIMEOUT: Duration = Duration::from_secs(3);
+const KEEPALIVE_WHILE_IDLE: bool = true;
+const MESSAGE_RECEIVER_CHANNEL_CAPACITY: usize = 100;
+
+#[derive(Debug, Clone)]
+pub enum ClientMessage {
+    ConnectionFailed(String),
+}
+
+/// Represents the state of a gRPC connection to the AMPD server.
+///
+/// This enum is used by the connection pool to communicate the connection status to clients.
+#[derive(Debug, Clone)]
+pub enum ConnectionState {
+    /// The connection is established and ready for service calls.
+    /// Contains the active gRPC transport channel.
+    Connected(transport::Channel),
+
+    /// The connection is not available and no reconnection attempt is in progress.
+    /// This is the initial state when the connection pool starts, or the final state after all reconnection attempts have failed.  
+    Disconnected,
```

Comment:
Why even make this state ambiguous? Just split it into unconnected and shut down (or keep the disconnected name for this one). Then it's clear that the handle should request a new connection for one, but not the other